spec/light-client/Lightclient.tla

```diff
@@ -1,111 +1,273 @@
+             (and pop the top element off the stack).
+             In contrast to the English spec, these requests are not processed immediately,
+             but one-by-one in a depth-first order (mind the optimization above).
+             *)
+            LET rightReq == [isLeft |-> FALSE, startHeight |-> (lh + rh) \div 2, endHeight |-> rh]
+                leftReq ==  [isLeft |-> TRUE, startHeight |-> lh, endHeight |-> (lh + rh) \div 2]
+            IN
+            <<TRUE, <<leftReq, rightReq>> \o Tail(requestStack)>>
+
+OnResponseHeader ==
+  /\ state = "working"
+  /\ inEvent.type = "responseHeader"
+  /\ storedHeaders' = storedHeaders \union { inEvent.hdr }
+  /\ LET res == OneStepOfBisection(storedHeaders')
+         verdict == res[1]
+         newStack == res[2]
+     IN
+      /\ requestStack' = newStack
+      /\ IF newStack = << >>
+         THEN /\ outEvent' = [type |-> "finished", verdict |-> verdict]
+              /\ state' = "finished"
+         ELSE /\ outEvent' = RequestHeaderForTopRequest(newStack)
+              /\ state' = "working"  
+
+LCInit ==
+    /\ state = "init"
+    /\ outEvent = NoEvent
+    /\ requestStack = <<>>
+    /\ storedHeaders = {}
+
+LCNext ==
+  OnStart \/ OnResponseHeader
+            
+            
+(********************* Lite client + Environment + Blockchain *******************)
+Init == BC!Init /\ EnvInit /\ LCInit
 
 Next ==
-  (* the blockchain may progress by adding one more block,
-     provided that the ultimate height has not been reached yet *)
-  /\ \/ height < ULTIMATE_HEIGHT /\ AppendBlock
-     \/ height = ULTIMATE_HEIGHT /\ UNCHANGED <<height, blockchain>> 
-  (* as time is passing, the minimal trusted height may increase *)
-  /\ minTrustedHeight' \in minTrustedHeight..ULTIMATE_HEIGHT
+  \/ LCNext  /\ UNCHANGED bcvars /\ UNCHANGED envvars
+  \/ EnvNext /\ UNCHANGED bcvars /\ UNCHANGED lcvars
+  \/ BC!Next /\ UNCHANGED lcvars /\ UNCHANGED envvars
+
+
+(* The properties to check *)
+\* check this property to get an example of a terminating light client
+NeverStart == state /= "working"
+
+NeverFinish == state /= "finished"
+
+
+\* Correctness states that all the obtained headers are exactly like in the blockchain.
+\* This formula is equivalent to Accuracy in the English spec, that is, A => B iff ~A => ~B.
+CorrectnessInv ==
+    outEvent.type = "finished" /\ outEvent.verdict = TRUE
+        => (\A hdr \in storedHeaders: hdr[1] = blockchain[hdr[1].height])
+
+PrecisionInv ==
+    outEvent.type = "finished" /\ outEvent.verdict = FALSE
+        => (\E hdr \in storedHeaders: hdr[1] /= blockchain[hdr[1].height])
+        
+\* TODO: add Termination that says that the light client always reaches the finished state
+\* under some conditions (when communicating with a correct full node).
+
+\* TODO: specify Completeness, which is a combination of Termination and Precision.
+\* TODO: Completeness in the English spec assumes that the lite client communicates
+\* with a correct full node.  
+
```

Comment:
We should probably also add equivalent of `Verification Condition` (English spec) property.